mobile-ui/src/main/java/org/sagebionetworks/research/mobile_ui/inject/ShowStepModule.java

```diff
@@ -32,62 +32,101 @@
+import org.sagebionetworks.research.presentation.model.implementations.FormUIStepViewBase;
+import org.sagebionetworks.research.presentation.model.implementations.UIStepViewBase;
+import org.sagebionetworks.research.presentation.model.interfaces.StepView;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-import dagger.Binds;
+import java.util.Map;
+
+import dagger.MapKey;
 import dagger.Module;
-import dagger.android.AndroidInjector;
-import dagger.android.support.FragmentKey;
+import dagger.Provides;
 import dagger.multibindings.IntoMap;
+import dagger.multibindings.Multibinds;
 
-@Module(subcomponents = {ShowStepFragmentSubcomponent.class, ShowCompletionStepFragmentSubcomponent.class,
-        ShowUIStepFragmentSubcomponent.class, ShowFormUIStepFragmentSubcomponent.class,
-        ShowActiveUIStepFragmentSubcomponent.class, ShowCountdownStepFragmentSubcomponent.class})
+// The ShowUIStepViewFragment needs the drawable module so it doesn't make sense to use one of these modules without
+// the other.
+@Module(includes = DrawableModule.class)
 public abstract class ShowStepModule {
-    @Binds
-    @IntoMap
-    @FragmentKey(ShowActiveUIStepFragment.class)
-    abstract AndroidInjector.Factory<? extends android.support.v4.app.Fragment>
-    bindShowActiveUIStepFragmentInjectorFactory(ShowActiveUIStepFragmentSubcomponent.Builder builder);
+    private static final Logger LOGGER = LoggerFactory.getLogger(ShowStepModule.class);
+
+    @MapKey
+    public @interface StepViewKey {
+        String value();
+    }
+
+    @Multibinds
+    abstract Map<String, ShowStepFragmentFactory> fragmentFactoryMap();
+
+    public interface ShowStepFragmentFactory {
+        @NonNull
+        ShowStepFragmentBase create(@NonNull StepView stepView);
+    }
+
+    @Provides
+    public static ShowStepFragmentFactory provideShowStepFragmentFactory(
+            Map<String, ShowStepFragmentFactory> showStepFragmentFactoryMap) {
+        return (@NonNull StepView stepView) -> {
+            if (showStepFragmentFactoryMap.containsKey(stepView.getType())) {
+                ShowStepFragmentBase fragment = showStepFragmentFactoryMap.get(stepView.getType()).create(stepView);
+                LOGGER.debug("Created fragment with class: {} from stepView: {}",
+                        fragment.getClass().getCanonicalName(), stepView);
+                if (fragment !=null ){
+                    return fragment;
+                }
+            }
+
+            // If we don't have a factory we default to the most general ShowStepFragment.
+            return ShowUIStepFragment.newInstance(stepView);
+        };
+    }
 
-    @Binds
+    @Provides
     @IntoMap
-    @FragmentKey(ShowGenericStepFragment.class)
-    abstract AndroidInjector.Factory<? extends android.support.v4.app.Fragment>
-    bindShowStepFragmentInjectorFactory(ShowStepFragmentSubcomponent.Builder builder);
+    @StepViewKey(ActiveUIStepViewBase.TYPE)
```

Comment:
This is an issue for a separate pr but I would like to see these divided into ActiveUIStepModule, UIStepModule, CompletionStepModule, etc like we did in mpower.

Reply:
I think they go together because you can't aren't really supposed to use mobile_ui without using all the base. It makes the most sense to split up in :sageresearch-app-sdk because those steps are optional building blocks for Bridge apps. 
Also note the change so the ShowStepFragment subcomponent includes now separated from the ShowStep modules because the ShowStep factories need to go above the PerformTaskFragment subcomponent's scope, while the injectors for the ShowStepFragments need to be within PerformTaskFragment's scope.